File: examples/cs/gmpub/gmpub.cs
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


 
/*! \file gmpub.cs
 *  This file contains a simple example publisher using the GMSEC.NET API.
 */



namespace gmpub
{
	using GMSEC.API;
    using GMSEC.API.UTIL;
	using GMSEC.API.Example.Common;
	using System;

	class gmpub : GmsecExample
	{
		private Config     config;
		private Connection conn;


		public gmpub()
		{
		}


		private void PrintUsage()
		{
			Log.Error("\nusage: gmpub_cs connectiontype=<middleware> " +
			          "[ <parameter>=<value> ]\n" +
			          "\n\tNote that the parameter 'connectiontype' is required. " +
			          "\n\tThe rest of other parameters are optional.\n" +
			          "\n\tserver=<server name> " +
			          "(required if middleware is not bolt/MB locally)" +
			          "\n\tsubject=<subject name>" +
			          "\n\titerations=<iterations>" +
			          "\n\tinterval_ms=<interval (milliseconds)>" +
			          "\n\tloglevel=<log level>" +
			          "\n\tcfgfile=<config_filename.xml>" +
			          "\n\nFor more information, see API User's Guide\n");
		}


        private void Publish(int iteration, String subject)
        {
            int value = iteration + 1;

            using (Message msg = new Message(subject, Message.MessageKind.PUBLISH))
			{
            	// Add Fields
				msg.AddField(new CharField("CHAR-FIELD", (sbyte) 'c'));

				msg.AddField(new BooleanField("BOOL-FIELD-TRUE", true));
				msg.AddField(new BooleanField("BOOL-FIELD-FALSE", false));

				msg.AddField(new I8Field("I8-FIELD",   (sbyte) value));
				msg.AddField(new I16Field("I16-FIELD", (Int16) value));
				msg.AddField(new I32Field("I32-FIELD",         value));
				msg.AddField(new I64Field("I64-FIELD", (Int64) value));

				msg.AddField(new U8Field("U8-FIELD",   (byte)   value));
				msg.AddField(new U16Field("U16-FIELD", (UInt16) value));
				msg.AddField(new U32Field("U32-FIELD", (UInt32) value));
				msg.AddField(new U64Field("U64-FIELD", (UInt64) value));

				msg.AddField(new F32Field("F32-FIELD", (Single)(1 + 1.0 / value)));
				msg.AddField(new F64Field("F64-FIELD", (Double)(1 + 1.0 / value)));

				msg.AddField(new StringField("STRING-FIELD", "This is a test"));

				msg.AddField(new BinaryField("BIN-FIELD", System.Text.Encoding.ASCII.GetBytes("JLMNOPQ")));

				msg.AddField(new I32Field("COUNT", iteration));

				try {
					if (config.GetBooleanValue("ENCRYPT"))
					{
						msg.AddField(new BooleanField("SEC-ENCRYPT", true));
					}
				}
				catch (GMSEC_Exception e) {
					Log.Debug(e.ToString());
				}

				// Publish Message
				conn.Publish(msg);

				// Display XML representation of the message
				Log.Info("Published GMSEC Message:\n" + msg.ToXML());
			}
        }


		private void Run(string[] args)
		{
			config = new Config(args);

			Initialize(config);
			AddToConfigFromFile(config);

			if (IsOptionInValid(config, args, false))
			{
				PrintUsage();
				return;
			}

			// Output GMSEC API version
			Log.Info(Connection.GetAPIVersion());

			try
			{
				// Create the connection
				conn = Connection.Create(config);

				// Connect
				conn.Connect();

				//Get Command Line Information
				String subject  = Get(config, "SUBJECT", "GMSEC.TEST.PUBLISH");
				int iterations  = Get(config, "ITERATIONS", 1);
				int interval_ms = Get(config, "INTERVAL_MS", 1000);
            
				// Output middleware version & Info
				Log.Info("Middleware version = " + conn.GetLibraryVersion());
				Log.Info("Using Subject '" + subject + "'");
				Log.Info("Publishing " + iterations + " message(s)");
            
				if (iterations > 1)
				{
					Log.Info("Publishing Interval "+interval_ms+" [ms]");
				}

				for (int i = 0; i < iterations; ++i)
				{
					Publish(i, subject);

					if (i < iterations - 1)
					{
						System.Threading.Thread.Sleep(interval_ms);
					}
				}
			}
			catch (GMSEC_Exception e)
			{
				Log.Error(e.ToString());
			}
			finally
			{
				if (conn != null)
				{
					// Disconnect
					conn.Disconnect();

					// Destroy the Connection
					Connection.Destroy(ref conn);
				}
			}
		}


		static void Main(string[] args)
		{
			new gmpub().Run(args);
		}
	}
}

File: examples/cpp/gmpub/gmpub.cpp
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/** 
 * @file gmpub.cpp
 * 
 * A C++ example demonstration of GMSEC publish/subscribe functionality.
 * The associated example gmsub.cpp will listen for this publish program.
 * gmsub should be run before gmpub.
 * 
 */

#include "../example.h"

#include <string>

using namespace gmsec::api;
using namespace gmsec::api::util;


class gmpub
{
public:
	gmpub(Config& config);
	~gmpub();
	bool run();

private:
	Config&     config;
	Connection* connection;

	void publish(int iteration, const std::string& subject);
};


gmpub::gmpub(Config &c)
	: config(c),
	  connection(0)
{
	/* Initialize config */
	example::initialize(c);
}


gmpub::~gmpub()
{
	if (connection)
	{
		connection->disconnect();

		Connection::destroy(connection);
	}

	Connection::shutdownAllMiddlewares();
}


bool gmpub::run()
{
	bool success = true;

	/* output GMSEC API version */
	GMSEC_INFO << Connection::getAPIVersion();

	try
	{
		//o Create the Connection
		connection = Connection::create(config);

		//o Connect
		connection->connect();

		//o Get information from the command line
		std::string subject     = example::get(config, "SUBJECT", "GMSEC.TEST.PUBLISH");
		int         iterations  = example::get(config, "ITERATIONS", 1);
		int         interval_ms = example::get(config, "INTERVAL_MS", 1000);

		//o Output middleware version
		GMSEC_INFO << "Middleware version = " << connection->getLibraryVersion();

		//o Output information
		GMSEC_INFO << "Using subject '" << subject.c_str() << "'";
		GMSEC_INFO << "Publishing " << iterations << " message(s)";

		if (iterations > 1)
		{
			GMSEC_INFO << "Publish interval " << interval_ms << " [ms]";
		}

		//o Publish message(s)
		for (int i = 0; i < iterations; ++i)
		{
			publish(i, subject);

			if (i < iterations - 1)
			{
				TimeUtil::millisleep(interval_ms);
			}
		}
	}
	catch (Exception& e)
	{
		GMSEC_ERROR << e.what();
		success = false;
	}

	return success;
}


void gmpub::publish(int iteration, const std::string& subject)
{
	int value = iteration + 1;

	GMSEC_INFO << "working";

	//o Create message
	Message message(subject.c_str(), Message::PUBLISH);

	//o Add fields to message
	message.addField("CHAR-FIELD", (GMSEC_CHAR) 'c');
	message.addField("BOOL-FIELD-TRUE", true);
	message.addField("BOOL-FIELD-FALSE", false);
	message.addField("I8-FIELD", (GMSEC_I8) value);
	message.addField("I16-FIELD", (GMSEC_I16) value);
	message.addField("I32-FIELD", (GMSEC_I32) value);
	message.addField("I64-FIELD", (GMSEC_I64) value);
	message.addField("U8-FIELD", (GMSEC_U8) value);
	message.addField("U16-FIELD", (GMSEC_U16) value);
	message.addField("U32-FIELD", (GMSEC_U32) value);
	message.addField("U64-FIELD", (GMSEC_U64) value);
	message.addField("COUNT", (GMSEC_I32) iteration);
	message.addField("STRING-FIELD", "This is a test");
	message.addField("F32-FIELD", GMSEC_F32(1 + 1. / value));
	message.addField("F64-FIELD", GMSEC_F64(1 + 1. / value));
	message.addField("BIN-FIELD", (GMSEC_BIN) "JLMNOPQ", 7);

	if (config.getBooleanValue("ENCRYPT", false))
	{
		message.addField("SEC-ENCRYPT", true);
	}

	//o Publish Message
	connection->publish(message);

	//o Display XML representation of message
	GMSEC_INFO << "Published:\n" << message.toXML();
}


int main(int argc, char* argv[])
{
	Config config(argc, argv);

	example::addToConfigFromFile(config);

	if (example::isOptionInvalid(config, argc))
	{
		example::printUsage("gmpub");
		return -1;
	}

	gmpub(config).run();
}
